### PR TITLE
Program iterator refactor

### LINKED ISSUE #67: Where to instantiate the Solver?
### Current implementation

Currently, the end user is responsible for instantiating the Solver and passing it to an iterator:

```julia
#main.jl
solver = Solver(grammar, :Int)
iter = BFSIterator(grammar, :Int, solver=solver, max_depth=4, max_size=8)
```

The iterator then passes the `max_depth` and `max_size` parameters to the solver:

```julia
function Base.iterate(iter::TopDownIterator)
    # Priority queue with number of nodes in the program
    pq :: PriorityQueue{State, Union{Real, Tuple{Vararg{Real}}}} = PriorityQueue()

    #TODO: these attributes should be part of the solver, not of the iterator
    solver = iter.solver
    solver.max_size = iter.max_size
    solver.max_depth = iter.max_depth

    enqueue!(pq, get_state(solver), priority_function(iter, get_grammar(solver), get_tree(solver), 0))
    return _find_next_complete_tree(iter.solver, pq, iter)
end
```

The iterator and solver are both holding the `Grammar`, but only the solver needs to store it. Since the iterator is require to hold the solver, it automatically has access to the grammar via `get_grammar(solver)`.

The iterator and solver are both holding the starting symbol, but again, only one of them actually needs to hold it. The starting symbol is a way to jump start the enumeration. In the current implementation, the top down iterator does not use the starting symbol anymore, because the responsibility of jumpstarting the search has been shifted to the solver.

Side note: the solver has two constructor signatures:
```julia
"""
    Solver(grammar::Grammar, sym::Symbol)

Constructs a new solver, with an initial state using starting symbol `sym`
"""
function Solver(grammar::Grammar, sym::Symbol; with_statistics=false)
    #...
end


"""
    Solver(grammar::Grammar, init_node::AbstractRuleNode)

Constructs a new solver, with an initial state of the provided [`AbstractRuleNode`](@ref).
"""
function Solver(grammar::Grammar, init_node::AbstractRuleNode; with_statistics=false)
    #...
end
```

### Target Implementation

The `solver` should be an optional keyworded argument of the program iterator constructor.
If the `solver` argument is not provided, the program iterator constructor should instantiate one.
Remove the `grammar`, `sym`, `max_depth` and `max_size` fields from the program iterator. (or ensure that the program iterator and the solver have the same values for these fields in the program iterator constructor)

### PR DESCRIPTION
**Pull request by nicolaefilat**
## @programiterator changes 

There are now 3 constructors for the iterator that all use the solver. This should close #67 

```jl
function iter(grammar, start_symbol, customargs..., max_depth = maxInt, max_size = maxInt, kwargs...)
    return (GenericSolver(grammar,start_symbol), customargs..., max_depth = max_depth, max_size = max_size,  kwargs...)
end
function iter(grammar, start_node, customargs..., max_depth = maxInt, max_size = maxInt, kwargs...)
    return (GenericSolver(grammar, start_node), customargs..., max_depth = max_depth, max_size = max_size,  kwargs....)
end
function iter(solver, customargs..., kwargs...)
    return (solver, customargs...,  ,  kwargs...)
end

```

Because users might by mistake add `solver` as a field to the function I have a check that disallows some _reserved name_ which are `solver`, `grammar`, `max_size` etc

## Comments 
The merge requests is quite big 😉because a lot of changes had to be done in all the places where iterators were created.
Hopefully from now one we are not going to change the macro again.

Nice todo would be to add tests that the @programiterator fails when users add `grammar` as their own field 

**Update**: The PR is merged (This PR depends on another PR (See [PR](https://github.com/Herb-AI/HerbConstraints.jl/pull/35)) from HerbConstraints that adds the `get_starting symbol` function)

## Documentation 
I just noticed that the stochastic functions have outdated documentation and I added a TODO there.
I could have fixed that in this PR but that would have been too many changes I think

I had to remove some of the @sourceCode4's tests because they were _too complex_ 😄

### CODE REVIEW

File: src/program_iterator.jl [hunk 1]
```diff
@@ -67,56 +67,112 @@ end
...
+            )
+        )
+        # this constructor should ONLY be used when there are kwarg fields 
+        # otherwise this will overwrite the default julia struct constructor
+        solver_constructor = Base.remove_linenums!(:(
+          # solver main constructor
+          function $(escaped_name)(solver::Solver, $(notkwargs...) ; $(kwargs_fields...) )
+              return $(escaped_name)(solver, $(field_names...))
+          end
+        ))
```
Comment (Whebon) [inline src/program_iterator.jl:117]: Does the following constructor also work?
```julia
solver = GenericSolver(grammar, :S, max_size=10, max_depth=5)
iterator = BFSIterator(solver)
```

**About the kwargs issue**: We could include `max_size` and `max_depth` as kwargs for the `solver_constructor` too. By default, the max_size and max_depth of the solver will be used. But if the user passes a specific max_size or max_depth to the iterator, they will be overwritten. For example:

```julia
solver = GenericSolver(grammar, :S, max_size=10, max_depth=5)
iterator = BFSIterator(solver, max_depth=3) #will overwrite solver.max_depth from 5 to 3. But keeps solver.max_size=10.
```

Comment (Whebon) [inline src/program_iterator.jl:117]:
```suggestion
        solver_constructor = Base.remove_linenums!(:(
          # solver main constructor
          function $(escaped_name)(solver::Solver, $(notkwargs...) ; max_size = nothing, max_depth = nothing, $(kwargs_fields...) )
              if !isnothing(max_size) solver.max_size = max_size end
              if !isnothing(max_depth) solver.max_depth = max_depth end
              return $(escaped_name)(solver, $(field_names...))
          end
        ))
```

Comment (nicolaefilat) [reply src/program_iterator.jl:117]: Ok I think it makes sense. I will commit this

File: src/genetic_search_iterator.jl [hunk 2]
```diff
@@ -137,13 +138,14 @@ Iterates the search space using a genetic algorithm. First generates a populatio
...
     validate_iterator(iter)
-    grammar = iter.grammar
+    grammar = get_grammar(iter.solver)
     
     population = Vector{RuleNode}(undef,iter.population_size)
 
     for i in 1:iter.population_size
         # sample a random nodes using start symbol and grammar
-        population[i] = rand(RuleNode, grammar, iter.sym, iter.maximum_initial_population_depth)
+        start_symbol = get_starting_symbol(iter.solver)
```
[the remark this answers is not in the document]
Comment (nicolaefilat) [reply src/genetic_search_iterator.jl:147]: That is a performance improvement

File: src/program_iterator.jl (hunk 1 above)
Comment (nicolaefilat) [reply src/program_iterator.jl]: The first one works nicely without issues as seen by this test
```jl
    @testset "Check default constructors with a solver" begin 
        solver = GenericSolver(g, :R, max_size=10, max_depth=5)
        iterator = BFSIterator(solver)
        @test get_grammar(iterator.solver) == g 
        @test get_max_size(iterator.solver) == 10 
        @test get_max_depth(iterator.solver) == 5
    end
 ```
 
 For the other one I had to make `solver` a kwarg in order to make julia figure out how to differentiate between the default constructor `Iterator(solver)` and `Iterator(solver; max_depth, max_size)`
 
```jl
    @testset "Check if max_depth and max_size are overwritten" begin 

        solver = GenericSolver(g, :R, max_size=10, max_depth=5)
        @test solver.max_size == 10
        @test solver.max_depth == 5
        # will overwrite solver.max_depth from 5 to 3. But keeps solver.max_size=10.
        iterator = BFSIterator(solver = solver, max_depth=3) # <- solver as kwarg
        @test get_max_size(solver) == 10 
        @test get_max_depth(solver) == 3
    end
    ```

**Comment by ReubenJ** [comment]
@nicolaefilat I've merged @Whebon's PRs now, and this introduced some conflicts. Could you update this branch/PR to resolve them? I'll review once it's up-to-date 👍

File: src/program_iterator_complicated.jl
Comment (ReubenJ) [inline src/program_iterator_complicated.jl:1]: Did you mean to keep/commit this file?

File: src/stochastic_functions/propose.jl [hunk 3]
```diff
@@ -28,12 +26,13 @@ The return function is a function that produces a list with all the subprograms
 # Arguments
 - `enumeration_depth::Int64`: the maximum enumeration depth.
 """
+# TODO: Refactor to not return functions
+# TODO: Update documentation with correct function signature
```
Comment (ReubenJ) [inline src/stochastic_functions/propose.jl:30]: Please add these (and other TODOs introduced) as GitHub issues and remove them from the source code—that way, we can keep track of progress/assign them to people as needed.

**Review by ReubenJ — CHANGES_REQUESTED** [review]

File: src/program_iterator_complicated.jl
Comment (nicolaefilat) [reply src/program_iterator_complicated.jl:1]: No I did not mean to commit that file. I will remove it

File: src/stochastic_functions/propose.jl (hunk 3 above)
Comment (nicolaefilat) [reply src/stochastic_functions/propose.jl:30]: Ok I will do that.

Comment (nicolaefilat) [reply src/stochastic_functions/propose.jl]: I added the issues #88 #87

File: test/test_programiterator_macro.jl [hunk 4]
```diff
@@ -70,43 +56,28 @@
...
-        @test it.max_depth == 1
-        @test it.intfield == 5
-        @test it.deffield == 6
-
-        it = ComplicatedIterator(g, :S, 5; max_depth=10)
-
-        @test it.max_depth == 10
-        @test length(it.grammar.rules) == 1
-        @test it.sym == :S
-        @test isnothing(it.deffield)
```
[the remark this answers is not in the document]
Comment (nicolaefilat) [reply test/test_programiterator_macro.jl:110]: The inner constructor tests do not pass. I talked with @sourceCode4, and he said that probably no-one will use the inner constructors in practice.

If users want to do define custom constructors, they can define the struct by themselves

[the remark this answers is not in the document]
Comment (ReubenJ) [reply test/test_programiterator_macro.jl:110]: Great, thanks! It's good to know that you two already discussed it. I'll merge after lunch 👍

**Review by ReubenJ — APPROVED** [review]

**Pull request merged**